lib/spack/spack/package.py

```diff
@@ -66,6 +66,40 @@
 """Allowed URL schemes for spack packages."""
 _ALLOWED_URL_SCHEMES = ["http", "https", "ftp", "file", "git"]
 
+import threading
+
+class DoInstallThread(threading.Thread):
+  """
+  This is a lightweight thread class to take care of executing
+  do_install calls in parallel.
+
+  To avoid the same package to be build multiple times,
+  the queued builds are being recorded. If a package thread finds
+  its target package already being in the queue, it just waits for
+  the other thread to finish.
+
+  """
+  _do_install_threads = {}
+  _queue_lock = threading.Lock()
+
+  def __init__(self, package, kwargs):
```

Comment:
@tgamblin - it did so for us for very deep stacks where some slow building (and in particular installing) packages are a bottleneck. We haven't gotten too many of those packages into spack yet, but I can provide you with spack build time measurements once done.

Doing only parallel fetches and serializing the rest I'd try once it's clear what the future of the staged packages in #543 is.